File: FrontEnd/Frontend_C#/SoftPac/SoftPacModel/FacturasDTO.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SoftPac.Model
{
    [Serializable]
    public class FacturasDTO : EliminableDTOBase
    {
        private int? facturaId;
        private String numeroFactura;
        private DateTime? fechaEmision;
        private DateTime? fechaRecepcion;
        private DateTime? fechaLimitePago;
        private String estado;
        private decimal montoTotal;
        private decimal montoIgv;
        private decimal montoRestante;
        private String regimenFiscal;
        private decimal tasaIva;
        private decimal otrosTributos;
        private AcreedoresDTO acreedor;
        private MonedasDTO moneda;
        private BindingList<DetallesFacturaDTO> detallesFactura;



        public FacturasDTO() : base()
        {
            FacturaId = null;
            NumeroFactura = null;
            FechaEmision = null;
            FechaRecepcion = null;
            FechaLimitePago = null;
            Estado = "I";
            MontoTotal = 0.0m;
            MontoIgv = 0.0m;
            MontoRestante = 0.0m;
            RegimenFiscal = null;
            TasaIva = 0.0m;
            OtrosTributos = 0.0m;
        }

        public FacturasDTO(int facturaId, String numeroFactura, DateTime fechaEmision,
                       DateTime fechaRecepcion, DateTime fechaLimitePago, String estado,
                       decimal montoTotal, decimal montoIgv, decimal montoRestante,
                       String regimenFiscal, decimal tasaIva, decimal otrosTributos,
                       AcreedoresDTO acreedor, MonedasDTO moneda)
        {
            this.FacturaId = facturaId;
            this.NumeroFactura = numeroFactura;
            this.FechaEmision = fechaEmision;
            this.FechaRecepcion = FechaRecepcion;
            this.FechaLimitePago = FechaLimitePago;
            this.Estado = estado;
            this.MontoTotal = montoTotal;
            this.MontoIgv = montoIgv;
            this.MontoRestante = montoRestante;
            this.RegimenFiscal = regimenFiscal;
            this.TasaIva = tasaIva;
            this.OtrosTributos = otrosTributos;
            this.Acreedor = acreedor;
            this.Moneda = moneda;
            this.DetallesFactura = new BindingList<DetallesFacturaDTO>();
        }

        public FacturasDTO(int factura_id, String numero_factura, DateTime fecha_emision,
                       DateTime fecha_recepcion, DateTime fecha_limite_pago, String estado,
                       decimal monto_total, decimal monto_igv, decimal monto_restante,
                       String regimen_fiscal, decimal tasa_iva, decimal otros_tributos,
                       AcreedoresDTO acreedor, MonedasDTO moneda, BindingList<DetallesFacturaDTO> detalles_factura)
        {
            this.FacturaId = factura_id;
            this.NumeroFactura = numero_factura;
            this.FechaEmision = fecha_emision;
            this.FechaRecepcion = fecha_recepcion;
            this.FechaLimitePago = fecha_limite_pago;
            this.Estado = estado;
            this.MontoTotal = monto_total;
            this.MontoIgv = monto_igv;
            this.MontoRestante = monto_restante;
            this.RegimenFiscal = regimen_fiscal;
            this.TasaIva = tasa_iva;
            this.OtrosTributos = otros_tributos;
            this.Acreedor = acreedor;
            this.Moneda = moneda;
            this.DetallesFactura = detalles_factura;
        }

        public FacturasDTO(FacturasDTO other) : base(other)
        {
            this.FacturaId = other.FacturaId;
            this.NumeroFactura = other.NumeroFactura;
            this.FechaEmision = other.FechaEmision;
            this.FechaRecepcion = other.FechaRecepcion;
            this.FechaLimitePago = other.FechaLimitePago;
            this.Estado = other.Estado;
            this.MontoTotal = other.MontoTotal;
            this.MontoIgv = other.MontoIgv;
            this.MontoRestante = other.MontoRestante;
            this.RegimenFiscal = other.RegimenFiscal;
            this.TasaIva = other.TasaIva;
            this.OtrosTributos = other.OtrosTributos;
            this.Acreedor = other.Acreedor;
            this.Moneda = other.Moneda;
            this.DetallesFactura = other.DetallesFactura;
        }


        public int? FacturaId { get => facturaId; set => facturaId = value; }
        public string NumeroFactura { get => numeroFactura; set => numeroFactura = value; }
        public DateTime? FechaEmision { get => fechaEmision; set => fechaEmision = value; }
        public DateTime? FechaRecepcion { get => fechaRecepcion; set => fechaRecepcion = value; }
        public DateTime? FechaLimitePago { get => fechaLimitePago; set => fechaLimitePago = value; }
        public string Estado { get => estado; set => estado = value; }
        public decimal MontoTotal { get => montoTotal; set => montoTotal = value; }
        public decimal MontoIgv { get => montoIgv; set => montoIgv = value; }
        public decimal MontoRestante { get => montoRestante; set => montoRestante = value; }
        public string RegimenFiscal { get => regimenFiscal; set => regimenFiscal = value; }
        public decimal TasaIva { get => tasaIva; set => tasaIva = value; }
        public decimal OtrosTributos { get => otrosTributos; set => otrosTributos = value; }
        public AcreedoresDTO Acreedor { get => acreedor; set => acreedor = value; }
        public MonedasDTO Moneda { get => moneda; set => moneda = value; }
        public BindingList<DetallesFacturaDTO> DetallesFactura { get => detallesFactura; set => detallesFactura = value; }


        public void AddDetalleFactura(DetallesFacturaDTO detalle)
        {
            if (DetallesFactura == null)
            {
                DetallesFactura = new BindingList<DetallesFacturaDTO>();
            }
            DetallesFactura.Add(detalle);
        }

    }
}

File: Backend_Java/SoftPacModel/src/main/java/pe/edu/pucp/softpac/model/FacturasDTO.java
package pe.edu.pucp.softpac.model;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Date;

public class FacturasDTO extends EliminableDTOBase{
    private Integer factura_id;
    private String numero_factura;
    private Date fecha_emision;
    private Date fecha_recepcion;
    private Date fecha_limite_pago;
    private String estado;
    private BigDecimal monto_total;
    private BigDecimal monto_igv;
    private BigDecimal monto_restante;
    private String regimen_fiscal;
    private BigDecimal tasa_iva;
    private BigDecimal otros_tributos;
    private AcreedoresDTO acreedor;
    private MonedasDTO moneda;
    private ArrayList<DetallesFacturaDTO> detalles_factura;

    public FacturasDTO() {
        super();
        factura_id=0;
        numero_factura=null;
        fecha_emision=null;
        fecha_recepcion=null;
        fecha_limite_pago=null;
        estado="I";
        monto_total=new BigDecimal("0.0");
        monto_igv=new BigDecimal("0.0");
        monto_restante=new BigDecimal("0.0");
        detalles_factura = new ArrayList<>();
    }

    public FacturasDTO(Integer factura_id, String numero_factura, Date fecha_emision,
                       Date fecha_recepcion, Date fecha_limite_pago, String estado,
                       BigDecimal monto_total, BigDecimal monto_igv, BigDecimal monto_restante,
                       String regimen_fiscal, BigDecimal tasa_iva, BigDecimal otros_tributos,
                       AcreedoresDTO acreedor, MonedasDTO moneda) {
        this.factura_id=factura_id;
        this.numero_factura = numero_factura;
        this.fecha_emision = fecha_emision;
        this.fecha_recepcion = fecha_recepcion;
        this.fecha_limite_pago = fecha_limite_pago;
        this.estado = estado;
        this.monto_total = monto_total;
        this.monto_igv = monto_igv;
        this.monto_restante = monto_restante;
        this.regimen_fiscal = regimen_fiscal;
        this.tasa_iva = tasa_iva;
        this.otros_tributos = otros_tributos;
        this.acreedor = acreedor;
        this.moneda = moneda;
        this.detalles_factura = new ArrayList<>();
    }

    public FacturasDTO(Integer factura_id, String numero_factura, Date fecha_emision,
                       Date fecha_recepcion, Date fecha_limite_pago, String estado,
                       BigDecimal monto_total, BigDecimal monto_igv, BigDecimal monto_restante,
                       String regimen_fiscal, BigDecimal tasa_iva, BigDecimal otros_tributos,
                       AcreedoresDTO acreedor, MonedasDTO moneda,  ArrayList<DetallesFacturaDTO> detalles_factura) {
        this.factura_id=factura_id;
        this.numero_factura = numero_factura;
        this.fecha_emision = fecha_emision;
        this.fecha_recepcion = fecha_recepcion;
        this.fecha_limite_pago = fecha_limite_pago;
        this.estado = estado;
        this.monto_total = monto_total;
        this.monto_igv = monto_igv;
        this.monto_restante = monto_restante;
        this.regimen_fiscal = regimen_fiscal;
        this.tasa_iva = tasa_iva;
        this.otros_tributos = otros_tributos;
        this.acreedor = acreedor;
        this.moneda = moneda;
        this.detalles_factura = detalles_factura;
    }

    public FacturasDTO(FacturasDTO factura) {
        super(factura);
        this.factura_id=factura.getFactura_id();
        this.numero_factura=factura.getNumero_factura();
        this.fecha_emision=factura.getFecha_emision();
        this.fecha_recepcion=factura.getFecha_recepcion();
        this.fecha_limite_pago=factura.getFecha_limite_pago();
        this.estado = factura.getEstado();
        this.monto_total=factura.getMonto_total();
        this.monto_igv=factura.getMonto_igv();
        this.monto_restante=factura.getMonto_restante();
        this.regimen_fiscal=factura.getRegimen_fiscal();
        this.tasa_iva=factura.getTasa_iva();
        this.otros_tributos=factura.getOtros_tributos();
        this.acreedor=factura.getAcreedor();
        this.moneda=factura.getMoneda();
        this.detalles_factura=factura.getDetalles_Factura();
    }

    public Integer getFactura_id() {
        return factura_id;
    }

    public void setFactura_id(Integer factura_id) {
        this.factura_id = factura_id;
    }

    public String getNumero_factura() {
        return numero_factura;
    }

    public void setNumero_factura(String numero_factura) {
        this.numero_factura = numero_factura;
    }

    public Date getFecha_emision() {
        return fecha_emision;
    }

    public void setFecha_emision(Date fecha_emision) {
        this.fecha_emision = fecha_emision;
    }

    public Date getFecha_recepcion() {
        return fecha_recepcion;
    }

    public void setFecha_recepcion(Date fecha_recepcion) {
        this.fecha_recepcion = fecha_recepcion;
    }

    public Date getFecha_limite_pago() {
        return fecha_limite_pago;
    }

    public void setFecha_limite_pago(Date fecha_limite_pago) {
        this.fecha_limite_pago = fecha_limite_pago;
    }

    public String getEstado() {
        return estado;
    }

    public void setEstado(String estado) {
        this.estado = estado;
    }

    public BigDecimal getMonto_total() {
        return monto_total;
    }

    public void setMonto_total(BigDecimal monto_total) {
        this.monto_total = monto_total;
    }

    public BigDecimal getMonto_igv() {
        return monto_igv;
    }

    public void setMonto_igv(BigDecimal monto_igv) {
        this.monto_igv = monto_igv;
    }

    public BigDecimal getMonto_restante() {
        return monto_restante;
    }

    public void setMonto_restante(BigDecimal monto_restante) {
        this.monto_restante = monto_restante;
    }

    public String getRegimen_fiscal() {
        return regimen_fiscal;
    }

    public void setRegimen_fiscal(String regimen_fiscal) {
        this.regimen_fiscal = regimen_fiscal;
    }

    public BigDecimal getTasa_iva() {
        return tasa_iva;
    }

    public void setTasa_iva(BigDecimal tasa_iva) {
        this.tasa_iva = tasa_iva;
    }

    public BigDecimal getOtros_tributos() {
        return otros_tributos;
    }

    public void setOtros_tributos(BigDecimal otros_tributos) {
        this.otros_tributos = otros_tributos;
    }

    public AcreedoresDTO getAcreedor() {
        return acreedor;
    }

    public void setAcreedor(AcreedoresDTO acreedor) {
        this.acreedor = acreedor;
    }

    public MonedasDTO getMoneda() {
        return moneda;
    }

    public void setMoneda(MonedasDTO moneda) {
        this.moneda = moneda;
    }
    
    public DetallesFacturaDTO getDetalle_Factura(Integer index) {
        return detalles_factura.get(index);
    }

    public void addDetalle_Factura(DetallesFacturaDTO tipo){
        if(detalles_factura==null){
            detalles_factura=new ArrayList<>();
        }
        this.detalles_factura.add(tipo);
    }

    public ArrayList<DetallesFacturaDTO> getDetalles_Factura(){
        return detalles_factura;
    }
    public void setDetalles_Factura(ArrayList<DetallesFacturaDTO> detalles_factura){
        this.detalles_factura=detalles_factura;
    }
}
